### PR TITLE
Avoid NullReferenceException if "params" property is null.

### PR DESCRIPTION
**Pull request by tintoy**
Specific fix for OmniSharp/csharp-language-server-protocol#24 (a more general fix will be offered as a separate PR).

### CODE REVIEW

File: src/JsonRpc/Reciever.cs [hunk 1]
```diff
@@ -77,7 +77,7 @@ protected virtual Renor GetRenor(JToken @object)
                 return new InvalidRequest(requestId, "Method not set");
             }
 
-            var hasParams = request.TryGetValue("params", out var @params);
+            var hasParams = request.TryGetValue("params", out var @params) && @params != null;
             if (hasParams && @params.Type != JTokenType.Array && @params.Type != JTokenType.Object)
```
Comment (david-driscoll) [inline src/JsonRpc/Reciever.cs]: We should change this line to better align to the spec.  [`params` if specified must be an array or object](http://www.jsonrpc.org/specification#parameter_structures).

Swap with:
`if (hasParams && @params?.Type != JTokenType.Array && @params?.Type != JTokenType.Object)`

File: src/JsonRpc/Reciever.cs [hunk 2]
```diff
@@ -54,7 +54,7 @@ protected virtual Renor GetRenor(JToken @object)
 
             object requestId = null;
             bool hasRequestId;
-            if (hasRequestId = request.TryGetValue("id", out var id))
+            if (hasRequestId = request.TryGetValue("id", out var id) && requestId != null)
```
[the remark this answers is not in the document]
Comment (tintoy) [reply src/JsonRpc/Reciever.cs]: In that case, shouldn't `hasRequestId` be `false` if `id == null`?

File: src/JsonRpc/Reciever.cs [hunk 3]
```diff
@@ -77,7 +77,7 @@ protected virtual Renor GetRenor(JToken @object)
                 return new InvalidRequest(requestId, "Method not set");
             }
 
-            var hasParams = request.TryGetValue("params", out var @params);
+            var hasParams = request.TryGetValue("params", out var @params) && @params != null;
```
[the remark this answers is not in the document]
Comment (tintoy) [reply src/JsonRpc/Reciever.cs]: But isn't this the source of the NRE? If `hasParams` is `true` but `@params` is `null`, then the line below will attempt to check `@params.Type` and crash, won't it?

File: src/JsonRpc/Reciever.cs (hunk 1 above)
Comment (tintoy) [reply src/JsonRpc/Reciever.cs]: Ah, ok, you want to use the `?.` operator. Sorry, fair enough - I missed that when viewing on my phone, I get you now. Will fix it up shortly :)

**Comment by tintoy** [comment]
> The thing that is causing this is probably a missing attribute in the client request object

The first place I ran into this was when troubleshooting an LPS4j-based client, but yeah - I could easily see various clients making a similar mistake.

File: src/JsonRpc/Reciever.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tintoy) [reply src/JsonRpc/Reciever.cs]: nvm - saw your other comment :)

**Comment by david-driscoll** [comment]
I wrote to follow the spec first (as per http://www.jsonrpc.org/specification) but if this comes up again with `params` being serialized as null (instead of absent), I think it's safe to just let it behave as a notification instead.  However it's kind of a grey area, which might cause some things to behave strangely...

**Comment by tintoy** [comment]
Thanks!

**Pull request merged**